### DISCUSSION TITLE
Evaluation guide for the grpc plugin

### DISCUSSION THREAD

**Discussion by pstorz** [bareos/bareos#2061] [Show and tell]
# Evaluation of the grpc plugin.
## Introduction
The bareos grpc-python plugin is a plugin that allows to run bareos python3 plugins in its own process. The plugin behaves like an intermediate piece that inserts itself between the bareos fileaemon plugin interface and the python plugin.

The grpc plugin starts a separate process, which loads the bareos filedaemon python plugin. Plugin calls that before went directly into the python plugin, now go into the grpc plugin, which transfers these calls via grpc into the plugin process which then hands them off to the python plugin that is loaded into the process.

## Prerequisites
To test the plugin, it is required to have a python plugin configured and working.

## Installation
Install the package bareos-filedaemon-grpc-python3-plugin.

> [!NOTE]
>
> Enterprise Linux Based Distributions require the epel and crb repositories to be activated:
> ```shell
> dnf install epel-release
> dnf config-manager --set-enabled crb
> ```

# Configuration
The usage of the grpc plugin is transparent when using existing python plugins. The behaviour of the
python plugin is exactly the same as without the grpc plugin.

To adapt an existing python plugin setup to use the grpc plugin mechanism is simple and requires the following steps:

Make sure the filedaemon loads the grpc plugin instead of the python3 plugin.
Change the filedaemon configuration in /etc/bareos/bareos-fd.d/client/myself.conf

   ```Plugin Names = "python3"``` 

needs to be changed to 

   ```Plugin Names = "grpc"```

Alter the `plugin =` configuration line so that the grpc plugin is called with the existing configuration line as parameter.
In the python plugin fileset configuration file, this is achieved by prefixing the existing line with `grpc:grpc-python-module:` like this:

An existing Plugin line
```
Plugin = "python3:module_name=bareos-fd-vmware..."
```
becomes
```
Plugin = "grpc:grpc-python-module:python3:module_name=bareos-fd-vmware..."
```
After these changes, the existing plugin job can be executed and will run as usual, the python interpreter will run in a separate process being controlled via grpc from the core.

> [!CAUTION]
>
> Jobs that were backed up with the python3 module cannot be recovered with the `grpc:grpc-python-module` and vice versa. This is a limitation that might be fixed in the future.

**Comment by bruno-at-bareos** [comment]
with such a system

```
 cat /etc/os-release
NAME="Red Hat Enterprise Linux"
VERSION="9.4 (Plow)"
```

after a successful subscription-manager register

```
dnf config-manager --set-enabled crb
Updating Subscription Management repositories.
Error: No matching repo to modify: crb.
```
and then of course still failing

```
dnf install bareos-filedaemon-grpc-python3-plugin
Updating Subscription Management repositories.
Red Hat Enterprise Linux 9 for x86_64 - AppStream (RPMs)                                                                                 10 MB/s |  52 MB     00:05
Red Hat Enterprise Linux 9 for x86_64 - BaseOS (RPMs)                                                                                    11 MB/s |  50 MB     00:04
Last metadata expiration check: 0:00:15 ago on Mon Mar 31 16:39:57 2025.
Error:
 Problem: cannot install the best candidate for the job
  - nothing provides libabsl_synchronization.so.2111.0.0()(64bit) needed by bareos-filedaemon-grpc-python3-plugin-24.0.2-1.el9.x86_64 from bareos
  - nothing provides libgrpc++.so.1.46()(64bit) needed by bareos-filedaemon-grpc-python3-plugin-24.0.2-1.el9.x86_64 from bareos
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)
```

**Reply by sduehr** [reply]
Both of the dependency errors refer to packages from EPEL (packages abseil-cpp and grpc-cpp), but there are also dependencies from CRB, which is required for EPEL generally.

As of https://docs.fedoraproject.org/en-US/epel/getting-started/ for RHEL 9 (assuming it is subscribed using subscription manager) this is how to add both CRB (CodeReady Builder) and EPEL:

```
subscription-manager repos --enable codeready-builder-for-rhel-9-$(arch)-rpms
dnf install https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm
```

**Reply by bruno-at-bareos** [reply]
Thanks for the link it has failed first try, but then has worked.

**Reply by arogge** [reply]
epel-release ships `/usr/bin/crb` which can be used to enable CRB cross-platform (i.e. `crb enable`)
